### ISSUE TITLE
routes.js {{yield}} equivalent?

### ISSUE THREAD

**Issue by nathanhammond** [dgeb/ember-engines#80]
Opening this issue as a reminder. Will come back and fill it out.

/cc @trentmwillis @nlfurniss

**Comment by krisselden** [comment]
@nathanhammond what is this?

**Comment by trentmwillis** [comment]
From an in person talk with Nathan, the idea is to implement some sort of construct to allow the concept of "yielding" inside an Engine's router map. As an example, if you wanted to define an "authentication" engine that defines login/logout routes, you would theoretically want an "authenticated" route to wrap other routes:

```js
export default buildRoutes(function() {
  this.route('login');
  this.route('logout');
  this.route('authenticated', function() {
    // yield in some routes here
  });
});
```

I'll fill in my personal opinions on this later, but that's the gist so far.

**Comment by nathanhammond** [comment]
A more complete straw man:

```javascript
// Host Application
Router.map(function() {
  this.mount('authentication-engine', { path: '/' }, function() {
    // This routes would be yielded
    this.route('foo');
    this.route('bar');
  });
});

// Engine
export default buildRoutes(function() {
  this.route('login');
  this.route('logout');
  this.route('authenticated', function() {
    this.yield('namedYield?');
  });
});
```

All of the yielded routes actually run in the parent app's `owner` context, which invalidates how we're handling setting of owners at this point for `outlet`s, but that is just implementation detail. If we choose to support named yields (slots?) then the API from the straw man would have to change. This should also probably support yielding engines.

> Yo dawg, I heard you like engines so I yielded an engine into your engine so you can compose your engines with engines.

**Comment by trentmwillis** [comment]
After thinking on this more, I think this would be a beneficial feature.

> All of the yielded routes actually run in the parent app's `owner` context

This would only really affect Engines, correct? If so, I think this makes sense. It distorts the mental model of the Engine/App tree a bit (by flattening it out), but I think we'd still have all the same functionality.

**Comment by nathanhammond** [comment]
> This would only really affect Engines, correct?

Correct, should only impact engines. Given the above straw man, `getOwner(fooRoute) => HostApplication`, even when `yield`ed in.

**Comment by rmachielse** [comment]
@nathanhammond @trentmwillis are you still interested in having something like this in ember-engines? This is exactly what we need for our application. Do you have any idea of the complexity of implementing this?

**Comment by trentmwillis** [comment]
I don't remember where we left off. I think we need some concrete use-cases to make a decision on whether it should be supported or not (likely also requires an RFC). That said, it'd like be pretty tricky to support, due to the context issues described above.

**Comment by Leooo** [comment]
We also have this use case on our side: different apps implementing appointment journeys, each of them have more or less the same steps in different orders (payment, questionnaire, selecting slots etc.), with different copies.

We would need basically the concept of a "container engine" to unify the appointment steps UX. Whereas current engines can well fit into a child <-> parent model relationship structure, it would be more to cater for a "is a" model where some of our engines would "extend" a generic appointment engine, if that makes sense.
